pypesto/objective/finite_difference.py

```diff
@@ -1,13 +1,215 @@
 """Finite differences."""
 
-from typing import List, Tuple, Union
+from typing import Callable, List, Tuple, Union
 import numpy as np
+import logging
 
 from .base import ObjectiveBase, ResultDict
 from .constants import (
     MODE_FUN, MODE_RES, FVAL, GRAD, HESS, RES, SRES,
 )
 
+logger = logging.getLogger(__name__)
+
+
+class Delta:
+    """Finite difference step size with automatic updating.
+
+    Parameters
+    ----------
+    delta:
+        (Initial) step size, either a float, or a vector of size (n_par,).
+        If not None, this is used as initial step size.
+    test_deltas:
+        Step sizes to try out in step size selection. If None, a range
+        [1e-2, 1e-3, ..., 1e-8] is considered.
+    update_condition:
+        A "good" step size may be a local property. Thus, this class allows
+        updating the step size if certain criteria are met, in the
+        :func:`pypesto.objective.finite_difference.Delta.update`
+        function.
+        Delta.CONSTANT means that the step size is only initially selected.
+        Delta.SPACE means that the step size is updated if the current
+        evaluation point is sufficiently far away from the last training point.
+        Delta.STEPS means that the step size is updated `max_steps` evaluations
+        after the last update.
```

Comment:
If `STEPS` are chosen, the `max_steps` is used to assess, whether or not Delta is recomputed. For `SPACE`, one would need an input such as `max_dist`.